### PR TITLE
Prod dimshuffle opt

### PR DESCRIPTION
**Pull request by mohammadpz**
I made a new PR regarding #399 and #3149 .
@carriepl , I need to talk to you once more to make sure that everything is ok and tests are expressive enough.

### CODE REVIEW

**Comment by carriepl** [comment]
There are currently conflicts so this PR will require a rebase.

**Comment by carriepl** [comment]
Also, according to Travis, some tests are failing with floatX=float32

File: theano/tensor/opt.py [hunk 1]
```diff
@@ -4076,37 +4079,49 @@ def local_sum_div_dimshuffle(node):
                                      'To disable this warning, set the Theano'
                                      ' flag warn.sum_div_dimshuffle_bug to'
                                      ' False.')
+                        break
```
Comment (carriepl) [inline theano/tensor/opt.py]: The change to this part will change the behavior in some cases

File: theano/tensor/tests/test_opt.py [hunk 2]
```diff
@@ -4987,6 +4988,77 @@ def test_local_sum_div_dimshuffle(self):
...
+        b_val = rng.randn(2).astype(config.floatX)
+        c_val = rng.randn(2, 2, 2).astype(config.floatX)
+        d_val = numpy.asarray(rng.randn(), config.floatX)
+
+        mode_with_opt = copy.copy(theano.compile.mode.get_default_mode())
+        mode_without_opt = copy.copy(theano.compile.mode.get_default_mode())
+        mode_with_opt._optimizer = mode_with_opt._optimizer.including(
+            'local_sum_prod_div_dimshuffle')
+        mode_without_opt._optimizer = mode_without_opt._optimizer.excluding(
+            'local_sum_prod_div_dimshuffle')
```
Comment (mohammadpz) [inline theano/tensor/tests/test_opt.py]: @carriepl , It seems that in both cases optimization is applied. Am I using ```including``` and ```excluding``` in a wrong way?

**Comment by carriepl** [comment]
The 'logical' tests currently fail in FAST_COMPILE. This is because, in this mode, elemwise fusion is not activated, hence there will be no composite elemwise. For the tests to pass, you can do one of the following : 
* always activate elemwise fusion in your mode_with_opt compilation mode
* always deactivate elemwise fusion and adapt expected outer operator to reflect that
* do nothing about the elemwise fusion and always accept what should be the outer operator with and without the fusion.

**Comment by carriepl** [comment]
Looks good. Thanks.

**Comment by mohammadpz** [comment]
@carriepl , Thanks for your kind assistance :+1:

**Pull request merged**